Document Sec-CH-Prefers-Color-Scheme Client Hint header

bershanskiy opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->
Document `Sec-CH-Save-Data` and `Sec-CH-Prefers-Color-Scheme` headers.

### Motivation
`Sec-CH-Save-Data` header shipped in Chromium 102 ([source](https://chromestatus.com/feature/5645928215085056)), `Sec-CH-Prefers-Reduced-Motion` will ship in Chromium 108 (see BCD).
<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details
Here is web.dev post: https://web.dev/user-preference-media-features-headers/
<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests
BCD PR (merged): https://github.com/mdn/browser-compat-data/pull/18203
<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

bershanskiy committed:
Document two Sec-CH-* Client Hint headers

hamishwillee commented:
Thanks @bershanskiy . There is significant overlap with the PR that @chrisdavidmills has done here: https://github.com/mdn/content/pull/22243.

I'd rather merge that because I have already reviewed it. Would it be OK if we did that and then integrate any useful differences that you have done?

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -0,0 +1,90 @@
+  </tbody>
+</table>
+
+The **`Sec-CH-Prefers-Reduced-Motion`** header expesses the user's media preference to see content with fewer animations and content shifts, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Reduced-Motion` to specify that the response is tailo ...
```
@chrisdavidmills The point about `vary` may be useful in your version. In particular you almost have to because I am pretty sure that if you set `Critical-CH` you _must_ also set `Vary`. That needs to be checked.

hamishwillee commented on files/en-us/web/svg/element/animate/index.md:
```diff
@@ -68,7 +68,7 @@ This element implements the {{domxref("SVGAnimateElement")}} interface.
 Blinking and flashing animation can be problematic for people with cognitive concerns such as Attention Deficit Hyperactivity Disorder (ADHD). Additionally, certain kinds of motion can be a trigger for Vestibular disorders, epilepsy, and migraine and Scotopic sensitivity.
 
-Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) to create a complimentary experience for users who have expressed a preference for no animated experiences.
+Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}} to create a complimentary experience for users who have expressed a preference for no ...
```
This kind of cross linking is really good.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,90 @@
+---
```
@chrisdavidmills A good starting point for this doc?

bershanskiy commented:
@hamishwillee Sure thing, I somehow missed #22243, thanks for prompt reply.

hamishwillee commented:
@bershanskiy So #22243 has now merged so you will have some conflicts. Can you please remove `Sec-CH-Prefers-Reduced-Motion` from the PR and then rebase against main? I'm hoping @chrisdavidmills can then review `Sec-CH-Prefers-Color-Scheme`.

bershanskiy commented:
> @bershanskiy So https://github.com/mdn/content/pull/22243 has now merged so you will have some conflicts. Can you please remove Sec-CH-Prefers-Reduced-Motion from the PR and then rebase against main?

Glad to hear this, I'm busy at the moment but I'll have some time over the weekend and will definitely rebase this by Monday.

bershanskiy committed:
f

bershanskiy committed:
Merge remote-tracking branch 'origin/main' into Sec-CH

bershanskiy committed:
Sec-CH-Save-Data

bershanskiy committed:
Standard template

bershanskiy commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -63,7 +64,7 @@ The server responds, telling the client via {{httpheader("Accept-CH")}} that it
 Content-Type: text/html
 Accept-CH: Sec-CH-Prefers-Reduced-Motion
-Vary: Sec-CH-Prefers-Color-Scheme
+Vary: Sec-CH-Prefers-Reduced-Motion
```
@chrisdavidmills I think this was a copy-paste error, considering line 71, but just wanted to check

chrisdavidmills replied:
Yup; see https://github.com/mdn/content/pull/22424

bershanskiy committed:
Merge remote-tracking branch 'origin/main' into Sec-CH

bershanskiy commented:
@hamishwillee I think this is ready for review. FYI, I added `Sec-CH-Save-Data` doc as well.

hamishwillee committed:
Merge branch 'main' into Sec-CH

hamishwillee commented on files/en-us/web/http/headers/save-data/index.md:
```diff
@@ -120,6 +120,7 @@ Content-Type: image/jpeg
 
 ## See also
 
+- {{HTTPHeader("Sec-CH-Save-Data")}} header which supercedes `Save-Data` and conveys the same information as a structured header
```
What do you mean by " as a structured header"

```suggestion
- {{HTTPHeader("Sec-CH-Save-Data")}} header which supercedes `Save-Data`
```

bershanskiy replied:
I just meant to write that `Save-Data` has value `on` (and has no other value) while `Sec-CH-Save-Data` has a boolean structured value written as `?0` or `?1` (meaning `on`). I do not mind dropping this bit of trivia, since people can deduce it by comparing the two articles.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -90,6 +91,7 @@ The client will include the header in subsequent requests in the current session
 ## See also
 
 - [Client hints](/en-US/docs/Web/HTTP/Client_hints)
+- `prefers-reduced-motion` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
```
```suggestion
- [`prefers-reduced-motion` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-reduced-motion/index.md:
```diff
@@ -6,6 +6,7 @@ tags:
   - Client hint
   - HTTP
   - HTTP Header
+  - prefers-reduced-motion
```
FYI only, the content guides still indicate that you should include tags like this in front matter. In fact they are pointless/ not used. The only tags that matter here are Experimental and Reference, as they affect the side bar.

hamishwillee committed:
Add Sec-CH-Save-Data to network client hints

hamishwillee commented on files/en-us/web/http/headers/index.md:
```diff
@@ -128,7 +132,9 @@ Network client hints allow a server to choose what information is sent based on
 - {{HTTPHeader("RTT")}}
   - : Application layer round trip time (RTT) in milliseconds, which includes the server processing time. This is part of the [Network Information API](/en-US/docs/Web/API/Network_Information_API).
-- {{HTTPHeader("Save-Data")}} {{experimental_inline}}
+- {{HTTPHeader("Sec-CH-Save-Data")}} {{experimental_inline}}
+  - : A boolean that indicates the user agent's preference for reduced data usage.
```
FYI added this.

hamishwillee replied:
```suggestion
- {{HTTPHeader("Sec-CH-Save-Data")}} {{experimental_inline}}
  - : A formatted value that indicates the user agent's preference for reduced data usage.
```

hamishwillee commented on files/en-us/web/http/headers/index.md:
```diff
@@ -104,18 +104,22 @@ The [UA client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints)
 ### Device client hints
 
-- {{HTTPHeader("Content-DPR")}} {{deprecated_inline}} {{experimental_inline}}
+- {{HTTPHeader("Content-DPR")}} {{deprecated_inline}}
```
FYI for all these, if something is deprecated the fact it was experimental is no longer considered relevant. So we just have deprecated.

bershanskiy committed:
f

bershanskiy commented:
@hamishwillee Thanks for promt reply, I fixed the remaining issues, could you take a look?

hamishwillee commented on files/en-us/web/api/animation/index.md:
```diff
@@ -92,7 +92,7 @@ You can see this in action in our simple [replace indefinite animations demo](ht
 Blinking and flashing animation can be problematic for people with cognitive concerns such as Attention Deficit Hyperactivity Disorder (ADHD). Additionally, certain kinds of motion can be a trigger for Vestibular disorders, epilepsy, and migraine, and Scotopic sensitivity.
 
-Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) to create a complimentary experience for users who have expressed a preference for no animated experiences.
+Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) and equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}} to create a complimentary experience for users who have expressed a preference for n ...
```
```suggestion
Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for no animated experiences.
```

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -179,6 +179,10 @@ Because media queries provide insights into the capabilities—and by extension,
 
+## User Agent Client Hints
+
+Some media queries have corresponding [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) which represent the values of discrete headers as HTTP response headers. These headers include: {{HTTPHeader("Sec-CH-Prefers-Color-Scheme")}}, {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}, and {{HTTPHeader("Sec-CH-Save-Data")}}.
```
```suggestion
Some media queries have corresponding [User Agent client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints), which are HTTP headers that request content that is already customized for the particular media requirement. These headers include: {{HTTPHeader("Sec-CH-Prefers-Color-Scheme")}}, {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}, and {{HTTPHeader("Sec-CH-Save-Data")}}.
```

hamishwillee commented on files/en-us/web/css/transition-timing-function/index.md:
```diff
@@ -104,7 +104,7 @@ transition-timing-function: unset;
 Some animations can be helpful such as to guide users to understand what actions are expected, to show relationships within the user interface, and to inform users as to what actions have occurred. Animations can help reduce cognitive load, prevent change blindness, and establish better recall in spatial relationships. However, some animations can be problematic for people with cognitive concerns ...
 
-Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) to create a complimentary experience for users who have expressed a preference for no animated experiences.
+Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) and equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}} to create a complimentary experience for users who have expressed a preference for n ...
```
```suggestion
Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for no animated experiences.
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented on files/en-us/web/http/headers/save-data/index.md:
```diff
@@ -120,6 +120,7 @@ Content-Type: image/jpeg
 
 ## See also
 
+- {{HTTPHeader("Sec-CH-Save-Data")}} header which supercedes `Save-Data`
```
```suggestion
- {{HTTPHeader("Sec-CH-Save-Data")}} (supercedes `Save-Data`)
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+HTTP/1.1 200 OK
+Content-Type: text/html
+Accept-CH: Sec-CH-Save-Data
+Vary: Sec-CH-Prefers-Color-Scheme
```
```suggestion
Vary: Sec-CH-Save-Data
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
````diff
@@ -0,0 +1,99 @@
+Host: example.com
+```
+
+The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Save-Data`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Save-Data` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints).
````
FWIW Would you actually make this Critical-CH in real life? It is ironic to apply the header as critical because you've already had the response that was not "low data" - so despite the user saying they wanted to save data you're now asking for the file again ... albeit in a lower data form.

I guess the argument for doing so is that the user won't see a significant change in quality on the next page they go to in the same session. I'm not completely convinced.

Anyway, I guess I'm saying "as a general rule", not every hint should be a critical hint, and not every example needs to demonstrate this. Something to think about.

bershanskiy replied:
Yes, I would consider making `Sec-CH-Save-Data` a critical hint and customizing page markup for low-data usage. For example [this blog](https://css-tricks.com/help-users-save-data/) writes about using `Save-Data` header and removing images, gifs and CSS styles from the markup entirely to save data. As far as I understand, back then there was no equivalent of `Critical-CH` so the browser would download the page twice (once requesting default full-featured experience and then immediately would reload requesting data-saving one). However, as far as I understand, `Critical-CH` enables server to avoid extra data transfers at the expense of extra round trips: if server sees a request without `Sec-CH-Save-Data`, it responds with `Accept-CH`, `Vary`, and `Critical-CH` and some "dummy" content (even just a spinner), then client sees `Critical-CH` and makes a new request with `Sec-CH-Save-Data` and server can now respond with customized markup.

It's possible to prepare three pre-baked responses (1) when header is not received (which uses media queries, including the upcoming `prefers-reduced-data`, image source sets, lazy loading, etc.), (2) data-saving mode (which removes almost everything, including unnecessary images) and (3) full experience which has lots of images defined in source sets, videos, media preloading, etc.

hamishwillee replied:
@bershanskiy I don't think it makes sense to do so - see my comment https://github.com/mdn/content/pull/22266/files#r1027381368

hamishwillee replied:
Depends what the spec says - it if is not low entropy, then that's fine.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+
+> **Note:** We've also specified `Sec-CH-Save-Data` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be present in t ...
+
+The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Save-Data` that it has a user preference for data-reduced-motion animations:
```
```suggestion
The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Save-Data` that it has a user preference for reduced data to be sent.
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`Sec-CH-Save-Data`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header indicates the user agent's preference for reduced data usage. This could be for reasons such as high transfer costs, slow connection speeds, etc.
```
```suggestion
The **`Sec-CH-Save-Data`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header indicates the user agent prefers less data be sent. This could be for reasons such as high transfer costs, slow connection speeds, and so on.
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`Sec-CH-Save-Data`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header indicates the user agent's preference for reduced data usage. This could be for reasons such as high transfer costs, slow connection speeds, etc.
+
```
I think this is a low entropy hint - you will need to check, and if so add some text similar to what is in `Save-Data` description, and also link to it from https://developer.mozilla.org/en-US/docs/Web/HTTP/Client_hints#low_entropy_hints

What makes this a little complicated is that you don't have to request a low entropy hint - if it is supported it gets sent by default. That means 
- you might not have the request cycle as below except to indicate that you want the value to vary.
- You wouldn't need the cycle below because it exists to ensure that the header gets sent before rendering - if it is low entropy that will always nappen.

Further, low entropy hints consume bandwidth - you wouldn't want to send both Save-Data and Sec-CH-Save-Data. So we need to find out how the spec handle this - are they both being sent?

hamishwillee committed:
Update files/en-us/web/http/headers/sec-ch-save-data/index.md

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+
+The **`Sec-CH-Save-Data`** [network client hint](/en-US/docs/Web/HTTP/Client_hints#network_client_hints) request header indicates the user agent prefers less data be sent. This could be for reasons such as high transfer costs, slow connection speeds, and so on.
+
+If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Save-Data`, the client can then respond with this header to indicate the user's preference for reduced data usage. The server can send the client appropriately adapted content, for example, JavaScript, videos, images, or CSS, to reduce the amount of data transfered for displaying subsequent rendered ...
```
1. Text here repeats itself
2. If this is a low entropy hint you don't need it in Accept-CH

```suggestion
On recieving this header in a request, a server can send the client appropriately adapted content.
For example, it can send fewer and lower resolution images and videos, and simpler CSS styles or HTML markup. 

Note that  `Sec-CH-Save-Data` should be used to reduce data sent to the client irrespective of the values of other client hints that affect network capability, such as {{HTTPHeader("Downlink")}} and {{HTTPHeader("RTT")}}.
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-save-data/index.md:
```diff
@@ -0,0 +1,99 @@
+
+- `<boolean>`
+
+  - : `?1` indicates that the user-agent prefers experience with reduced data usage (`true`).
+    `?0` indicates that user-agent prefers a full-featured experience, potentially with data-intensive components (`false`).
```
```suggestion
  - : `?1` indicates that the user-agent prefers that less data be sent (`true`).
    `?0` indicates that the user-agent prefers the "normal data" version of the site (`false`).
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,102 @@
+
+{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}
+
+The **`Sec-CH-Prefers-Color-Scheme`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (e.g. light or dark mode) or a user agent setting.
```
```suggestion
The **`Sec-CH-Prefers-Color-Scheme`** [user preference media feature client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (for example, light or dark mode) or a user agent setting.
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
````diff
@@ -0,0 +1,102 @@
+Host: example.com
+```
+
+The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Color-Scheme`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Prefers-Color-Scheme` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints).
````
FYI I think it makes sense to use Critical-CH in this example (as you have done) because the user has very clearly inidicated this preference, and you'll get a jarring disconnect if you navigate to another page on the site.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,102 @@
+
+- `<preference>`
+
+  - : The user agent's preference for dark or light content. This is often taken from the underlying operating system's setting. The value of this directive can be either `light` or `dark`. A string indicating the user preference, either `"light"` or `"dark"`.
```
I have a preference for stating the values and then the source.

```suggestion
  - : A string indicating the user agent's preference for dark or light content: `"light"` or `"dark"`.
    The value may originate from a corresponding setting in the underlying operating system.
```

bershanskiy commented:
Also, the bot pointed out that some BCD queries are failing because the data is missing. This is partially fixed by already-released [BCD 5.2.17](https://github.com/mdn/browser-compat-data/releases/tag/v5.2.17) and partially by in-progress PR https://github.com/mdn/browser-compat-data/pull/18243

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
````diff
@@ -0,0 +1,102 @@
+Sec-CH-Prefers-Color-Scheme: "dark"
+```
+
+The client will include the header in subsequent requests in the current session unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.
````
```suggestion
The client will include the header in subsequent requests in the current session, unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.
```

hamishwillee committed:
Apply suggestions from code review

bershanskiy commented:
Thanks for thorough review, I probably won't have time to address all points today, but I'll make sure to do it tomorrow.

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented:
@bershanskiy Tomorrow is fine. I've merged some of my minor comments. The big problem is sec-ch-save-data - if this is a low entropy hint then it is sent all the time - in which case making it a critical hint is pointless. We need to establish how that is handled, and what is happening to Save-Data (i.e. does it remain critical too, even though superseded).

https://github.com/mdn/content/pull/22266/files#r1027381368

chrisdavidmills commented:
@bershanskiy Hi there! Are you going to have time to return to this one soon? I'm only asking because I am about to start working on `Critical-CH`, and I remembered you had a PR open for this one.

Let me know if you need any help getting this one finished. Thanks!

bershanskiy commented:
@chrisdavidmills I'll have time to finish this PR during the weekend, if that works for you. Turned out that I was wrong abut `Save-Data` header (I got confused from reading source code and documentation, so I asked spec author and he fixed the Chromestatus entry). Also, I'm going to drop `Critical-CH` reference from "save data" page and leave onlythe  `Vary` suggestion.

bershanskiy committed:
Merge branch 'main' into Sec-CH

bershanskiy committed:
Merge remote-tracking branch 'origin/Sec-CH' into Sec-CH

bershanskiy committed:
f

bershanskiy committed:
f

bershanskiy committed:
f

bershanskiy committed:
f

bershanskiy committed:
f

bershanskiy commented:
@chrisdavidmills I finalized the PR, in short: the original version claimed that the `Save-Data` was renamed to `Sec-CH-Prefers-Reduced-Data`, repeating the Chrome Status entry. This was not correct: the name and value did not change, the header just got a Permissions Policy controls.

hamishwillee commented:
@chrisdavidmills When you've done your round of review, please ping me if you want a final sanity check.

chrisdavidmills commented on files/en-us/web/css/@media/index.md:
```diff
@@ -183,6 +183,10 @@ Because media queries provide insights into the capabilities—and by extension,
 
+## User Agent Client Hints
+
+Some media queries have corresponding [User Agent client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints), which are HTTP headers that request content that is already customized for the particular media requirement. These headers include: {{HTTPHeader("Sec-CH-Prefers-Color-Scheme")}}, {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}, and {{HTTPHeader("Save-Data")}}.
```
Bit of a nit, but you could remove "that is already" from this sentence to tighten it up a bit.

hamishwillee replied:
```suggestion
Some media queries have corresponding [User Agent client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints), which are HTTP headers that request content that are customized for the particular media requirement. These headers include: {{HTTPHeader("Sec-CH-Prefers-Color-Scheme")}}, {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}, and {{HTTPHeader("Save-Data")}}.
```

chrisdavidmills commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,103 @@
+
+The **`Sec-CH-Prefers-Color-Scheme`** [user preference media feature client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (for example, light or dark mode) or a user agent setting.
+
+If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Color-Scheme`, the client can then respond with this header to indicate the user's preference for a specific color scheme. The server can send the client appropriately adapted content, for example, images or CSS, to display light text on dark background on subsequent rendered content.
```
The last sentence here is a bit awkward to read because of all the commas. How about "The server can send the client appropriately adapted content including images or CSS to display a light or dark mode for subsequent rendered content." ?

hamishwillee replied:
Yes, that is more readable.

```suggestion
If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Color-Scheme`, the client can then respond with this header to indicate the user's preference for a specific color scheme. The server can send the client appropriately adapted content including images or CSS to display a light or dark mode for subsequent rendered content.
```

chrisdavidmills commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,103 @@
+  </tbody>
+</table>
+
+The **`Sec-CH-Prefers-Color-Scheme`** header expesses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a part ...
```
* "If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header" -> "If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header"
* "Value reported by" -> "The value reported by"
* "used" -> "used"

hamishwillee replied:
Integrates Chris suggestions. I tend to split these up paragraph into per-line sentences - but that is my taste - makes reviewing easier.

```suggestion
The **`Sec-CH-Prefers-Color-Scheme`** header expresses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particular color scheme.
The value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value used in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.
```

chrisdavidmills commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,103 @@
+
+The **`Sec-CH-Prefers-Color-Scheme`** header expesses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a part ...
+
+`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theiry be used for user fingerprinting.
```
"theiry" -> "theory"

hamishwillee replied:
@chrisdavidmills FWIW I just fix the obvious typos and merge them. Saves a little round tripping.
```suggestion
`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theory be used for user fingerprinting.
```

chrisdavidmills commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,103 @@
+## See also
+
+- [Client hints](/en-US/docs/Web/HTTP/Client_hints)
+- `prefers-color-scheme` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-color-scheme)
```
Missing an opening square bracket here.

hamishwillee replied:
FWIW fix as suggestion and self merge if there is "no question" it is just a bug. This is the "mike smith" review method :-)

```suggestion
- [`prefers-color-scheme` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-color-scheme)
```

chrisdavidmills reviewed (commented):
@bershanskiy looking really good. I've commented on a few small bits, which should be really quick to fix.

@hamishwillee I'll need you to do the final approve and merge. Cheers!

hamishwillee committed:
Chris typo fixes

hamishwillee committed:
Add link in accessibility section

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -175,14 +175,19 @@ To best accommodate people who adjust a site's text size, use [`em`](/en-US/docs
 Both [`em`](/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#numeric_values) and [`px`](/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#numeric_values) are valid units, but [`em`](/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#numeric_values) works better if the user changes the browser text size.
 
-Also consider using Level 4 media queries to improve the user's experience. For example, `prefers-reduced-motion` to [detect if the user has requested that the system minimize the amount of animation](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) or motion it uses.
+Also consider media queries or [HTTP user agent client hints](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) to improve the user's experience.
```
FYI, I added this.

chrisdavidmills replied:
Looks good

hamishwillee commented on files/en-us/web/css/transition-timing-function/index.md:
```diff
@@ -104,7 +104,7 @@ transition-timing-function: unset;
 Some animations can be helpful such as to guide users to understand what actions are expected, to show relationships within the user interface, and to inform users as to what actions have occurred. Animations can help reduce cognitive load, prevent change blindness, and establish better recall in spatial relationships. However, some animations can be problematic for people with cognitive concerns ...
 
-Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) to create a complimentary experience for users who have expressed a preference for no animated experiences.
+Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for  ...
```
```suggestion
Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for less animation.
```

hamishwillee commented on files/en-us/web/html/element/portal/index.md:
```diff
@@ -48,7 +48,7 @@ Portals are given a default label which is the title of the embedded page. If no
 Portals used for prerendering only should be hidden with the hidden HTML attribute or the CSS {{cssxref("display")}} property with a value of `none`.
 
-When using animations during portal activation the {{cssxref("@media/prefers-reduced-motion", "prefers-reduced-motion")}} [media feature](/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#media_features) should be respected.
+When using animations during portal activation the {{cssxref("@media/prefers-reduced-motion", "prefers-reduced-motion")}} [media feature](/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#media_features) and equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}} should be respected.
```
```suggestion
When using animations during portal activation, the {{cssxref("@media/prefers-reduced-motion", "prefers-reduced-motion")}} [media feature](/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#media_features) and equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}} should be respected.
```

hamishwillee committed:
Update files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,104 @@
+The **`Sec-CH-Prefers-Color-Scheme`** header expresses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particul ...
+The value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value used in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.
+
+`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theory be used for user fingerprinting.
```
```suggestion
`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. A user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy, since the user's preference could in theory be used for fingerprinting.
```

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
```diff
@@ -0,0 +1,104 @@
+  </tbody>
+</table>
+
+The **`Sec-CH-Prefers-Color-Scheme`** header expresses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particul ...
```
First sentence feels repetitive - i.e we already said this. Also seems a bit odd to have some content below the table and some above it. Usually if there is too much to say above the table then there is a separate "Description" section for the detail, and the information above the table is kept very succint.

chrisdavidmills replied:
In my most recent commit, I updated this section to make it less repetitive and be more about using it over the media query version if performance is a critical consideration in this context, and what other headers are needed alongside it. I also placed it under a "Usage notes" heading.

hamishwillee commented on files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md:
````diff
@@ -0,0 +1,104 @@
+Critical-CH: Sec-CH-Prefers-Color-Scheme
+```
+
+> **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be p ...
````
```suggestion
> **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate that responses should be separately cached based on the value of this header (even if the URL stays the same). 
> Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.
```

hamishwillee commented:
@chrisdavidmills @bershanskiy I've integrated Chris' feedback and added a few more suggestions and comments. It is very close. I'm more than happy to approve once those have been looked at.

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -175,14 +175,19 @@ To best accommodate people who adjust a site's text size, use [`em`](/en-US/docs
 
 Because of this potential, a browser may opt to fudge the returned values in some manner in order to prevent them from being used to precisely identify a computer. A browser might also offer additional measures in this area; for example, if Firefox's "Resist Fingerprinting" setting is enabled, many media queries report default values rather than values representing the actual device state.
 
+## User Agent Client Hints
```
Actually not entirely sure about this whole section - its not like security or accessibility, which are considerations about THIS API.

Maybe this could be a level 3 heading under description? Not sure. It is worth mentioning.

chrisdavidmills replied:
Moved in my latest commit, as suggested

bsmth commented:
Hi @bershanskiy - thanks a lot for the submission. There's some review comments still outstanding, do you plan to come back to this one?

chrisdavidmills committed:
Update files/en-us/web/css/@media/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/css/transition-timing-function/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/html/element/portal/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

chrisdavidmills committed:
Move user agent client hints section

As suggested by Hamish

chrisdavidmills reviewed (approved):
@hamishwillee I've decided to take over this PR, as it has been dormant for long enough. I have committed your suggestions and made some changes to satisfy other comments, and I think it looks OK. Let me know if there is anything else you think this needs.

chrisdavidmills committed:
Merge branch 'main' into Sec-CH

chrisdavidmills committed:
Update files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Reduce repetition of text following table

bsmth committed:
fix(API): Missing newline at EOF

hamishwillee commented on files/en-us/web/api/animation/index.md:
```diff
@@ -73,7 +73,7 @@ The **`Animation`** interface of the [Web Animations API](/en-US/docs/Web/API/We
 Blinking and flashing animation can be problematic for people with cognitive concerns such as Attention Deficit Hyperactivity Disorder (ADHD). Additionally, certain kinds of motion can be a trigger for Vestibular disorders, epilepsy, and migraine, and Scotopic sensitivity.
 
-Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) to create a complimentary experience for users who have expressed a preference for no animated experiences.
+Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [User Agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for  ...
```
Capitalization concerns me - either "User agent" to match linked topic or none at all.
```suggestion
Consider providing a mechanism for pausing or disabling animation, as well as using the [Reduced Motion Media Query](/en-US/docs/Web/CSS/@media/prefers-reduced-motion) (or equivalent [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}) to create a complimentary experience for users who have expressed a preference for no animated experiences.
```

chrisdavidmills replied:
Yup, I'd agree with that.

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -163,13 +163,20 @@ You can also combine multiple media queries into a single rule by separating the
 - `or`
   - : Equivalent to the `,` operator. Added in Media Queries Level 4.
 
+### User Agent Client Hints
```
IMO we don't treat User Agent as a proper noun anywhere else in MDN right. The only place this should be capitalized is  if we are referring specifically/in-full to the "User Agent Client Hints API", which is a proper name. 

```suggestion
### User agent client hints
```

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -163,13 +163,20 @@ You can also combine multiple media queries into a single rule by separating the
 
+### User Agent Client Hints
+
+Some media queries have corresponding [User Agent client hints](/en-US/docs/Web/HTTP/Client_hints).
```
```suggestion
Some media queries have corresponding [user agent client hints](/en-US/docs/Web/HTTP/Client_hints).
```

hamishwillee commented on files/en-us/web/css/@media/index.md:
```diff
@@ -163,13 +163,20 @@ You can also combine multiple media queries into a single rule by separating the
+
+Some media queries have corresponding [User Agent client hints](/en-US/docs/Web/HTTP/Client_hints).
+These are HTTP headers that request content that is pre-optimized for the particular media requirement.
+They include: {{HTTPHeader("Sec-CH-Prefers-Color-Scheme")}}, {{HTTPHeader("Sec-CH-Prefers-Reduced-Motion")}}, and {{HTTPHeader("Save-Data")}}.
```
`Save-Data` is not a user-agent client hint and it is implied that it is by the first link in this paragraph. It used to be a network client hint, and now it is is "unclear" - it recently moved into the save-data API. Network client hint is still the closest grouping.

Perhaps just swap to another hint that has a media query?

chrisdavidmills replied:
I've just removed `Save-Data`, reducing the list down to two, as this paragraph seems to be talking explicitly about user agent client hints that specific mirror media query functionality. And with this, PR, we'll only have those two documented ;-)

hamishwillee committed:
Remove duplicate Sec-CH-Prefers-Reduced-Motion

This removes the original, because I like these two items together at the end, and the new description is better.

hamishwillee commented:
@chrisdavidmills Thanks very much for taking this over. Just a couple of nits - happy to merge after you have looked at those.

chrisdavidmills committed:
Update files/en-us/web/api/animation/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/css/@media/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update files/en-us/web/css/@media/index.md

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

chrisdavidmills committed:
Update listed ua client hints

chrisdavidmills reviewed (commented):
OK, I think that's all your comments answered; thanks for the review @hamishwillee. Feel free to merge away if you are happy.

hamishwillee commented:
THanks. Good to get this in. Also to @bershanskiy For kicking it off in the first place.
